DiffCallback for DrawerLayoutAdapter

Vafeen opened the pull request:
Why notifyDataSetChanged() is Suboptimal:
Full Redraw: Calling notifyDataSetChanged() triggers a complete redraw of the entire RecyclerView, even if only a single item has changed.
Resource Intensive: This demands significant computational resources, particularly for large lists, as all items must be re-bound and re-drawn.
Undesirable Animations: A full redraw resets any animations associated with list items, leading to a visually jarring user experience.
Why DiffUtil is More Optimal:
Precise Updates: DiffUtil calculates the minimal set of changes (additions, removals, changes, moves) required to transform the old list into the new list.
Targeted Operations: Based on these changes, the adapter uses notifyItemInserted(), notifyItemRemoved(), notifyItemChanged(), and notifyItemMoved() to update only the items that actually need updating.
Improved Performance: This significantly reduces the load on the system, resulting in a smoother and more responsive UI, especially with large lists or frequent data updates.
Preservation of Animations: DiffUtil preserves RecyclerView item animations, providing a more polished visual experience.
In summary, DiffUtil provides granular control over RecyclerView updates, minimizing the amount of work needed to refresh displayed data, in contrast to the all-or-nothing approach of notifyDataSetChanged().

Vafeen commented:
@Arash63Dehghan 
@DrKLO 
@Bubu 
@alabiaga 
Hi, will it be merged?

yashikada commented:
> will it be merged?

I think no, is more important make changes to emoi or similar feature, maybe some colors.

Vafeen commented:
> I think no, is more important make changes to emoi or similar feature, maybe some colors.

It's not a feature, notifyDataSetChanged calls redrawing more often than necessary and i wanna fix it

yashikada commented:
I tested your PR and don't work very well, these import are useless:

```
import androidx.annotation.Nullable;
import org.jetbrains.annotations.NotNull;
```

if I switch account the blue flag is not updated, so I see the previous account as selected and not the current.
I don't approve this PR.

vafeen commented:
PR Description

🔄 Major Adapter Refactoring with DiffUtil Implementation

This PR completely refactors the DrawerLayoutAdapter to implement a modern, efficient architecture using DiffUtil for optimal RecyclerView performance.

🚀 Key Changes

1. Unified Data Source Architecture
- Replaced multiple scattered data structures with a single List<AbstractItem> as the single source of truth
- Implemented proper item typing with AbstractItem base class and specific implementations:
  - ProfileItem, EmptyItem, DividerItem, AccountItem, ActionItem, AddAccountItem

2. DiffUtil Integration
- Added FullDrawerDiffCallback for efficient item diffing
- Replaced notifyDataSetChanged() with redrawAdapterData() method
- Performance benefits: Only changed items are animated, not the entire list

3. Cleaner Position Handling
- Eliminated complex position calculations (position -= 2, getAccountRowsCount())
- Simplified getItemViewType() and onBindViewHolder() logic
- Fixed position-related bugs in account swapping and click handling

4. Enhanced Type Safety
- Replaced magic numbers with proper constants (VIEW_TYPE_PROFILE, VIEW_TYPE_ACTION, etc.)
- Added proper type checking in isEnabled() method

📊 Performance Improvements

- Before: notifyDataSetChanged() caused complete list redraw
- After: redrawAdapterData() with DiffUtil calculates minimal changes
- Result: Smoother animations, better performance, especially when toggling accounts

🛠 API Changes

Deprecated:
`@Deprecated(since = "12.1.1", forRemoval = true)
@Override
public void notifyDataSetChanged()`

New Method:
`public void redrawAdapterData()`

🔧 Updated Usage

All callers have been updated to use the new method:
- LaunchActivity - multiple locations
- Internal adapter methods
- Notification handlers

🐛 Bug Fixes

- Fixed account swapping position calculation
- Improved click handling reliability  
- Resolved potential index out of bounds issues
- Better handling of account section visibility toggles

✅ Testing

- Verified account switching works correctly
- Tested account section expand/collapse animations
- Confirmed all menu items display properly
- Validated click handlers work as expected

📝 Notes

The refactor maintains full backward compatibility while providing significant performance improvements. The deprecated notifyDataSetChanged() method automatically redirects to the new implementation.

ausboymax commented:
I'm sorry, I misunderstood it.

ในวันที่ พ. 29 ต.ค. 2025 08:58 A ***@***.***> เขียนว่า:

> *vafeen* left a comment (DrKLO/Telegram#1842)
> <https://github.com/DrKLO/Telegram/pull/1842#issuecomment-3459316496>
>
> Oh yeah, I can understand you. I was being ironic, unfortunately lately tg
> only thinks about the graphic aspect and not the rest. For example they
> updated boringssl only recently from over 2 years ago. They still compile
> with ndk 21 when now there is version 28 that maybe improves the
> performance.
>
> I understand that now tg only thinks about the graphic aspect, but i am
> trying to improve performance, because if we neglect it always, app won't
> work as fast as now. In one place we lost some hundred milliseconds, in
> other, in third, and in the result the user experience fell.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DrKLO/Telegram/pull/1842#issuecomment-3459316496>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2XGEBVVRJQETNZONRTJ3D32ANNPAVCNFSM6AAAAABWW3B3WSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTINJZGMYTMNBZGY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
